File: HavenApp/HavenApp/Models/FollowingSnapshot.swift
import Foundation

// MARK: - Relay-Fetched Kind 3 Events (Recovery)

/// Represents a Kind 3 contact list event fetched from a relay.
/// Used by the recovery UI to display historical contact lists.
struct Kind3Event: Identifiable {
    let id: String          // Nostr event ID
    let createdAt: Date     // From created_at timestamp
    let pubkeys: [String]   // Hex pubkeys extracted from p-tags
    let pTags: [[String]]   // Full NIP-02 p-tags (preserves relay hints, petnames)
    let content: String     // Kind 3 content field (relay config JSON)
    var followCount: Int { pubkeys.count }
}

// MARK: - Local Snapshots (Automatic Backup)

/// A point-in-time capture of the user's contact list, saved locally.
struct FollowingSnapshot: Codable, Identifiable {
    let id: UUID
    let capturedAt: Date
    let pubkeys: [String]
    let pTags: [[String]]
    let contactListContent: String
    /// `created_at` of the kind-3 event this snapshot represents. Optional so
    /// legacy snapshots (written before this field existed) still decode rather
    /// than wiping the durable backup. Used to refuse older relay copies that
    /// would clobber a newer local edit.
    var contactListCreatedAt: Int64?
    var followCount: Int { pubkeys.count }
}

/// Container for all snapshots belonging to one account.
struct FollowingSnapshotStore: Codable {
    var snapshots: [FollowingSnapshot]
    static let maxSnapshots = 50
}

File: NostrVault/app/src/main/java/com/nostrvault/data/model/FollowingSnapshot.kt
package com.nostrvault.data.model

import kotlinx.serialization.Serializable

/**
 * A snapshot of the user's following list at a point in time.
 * Used for following backup/restore functionality.
 */
@Serializable
data class FollowingSnapshot(
    val id: String,
    val capturedAt: Long,
    val pubkeys: List<String>,
    val pTags: List<List<String>>,
    val contactListContent: String,
    /**
     * created_at (unix seconds) of the kind-3 event this snapshot represents.
     * Null for legacy snapshots written before this field existed. Used to refuse
     * older relay copies that would clobber a newer local edit.
     */
    val contactListCreatedAt: Long? = null,
) {
    val followCount: Int get() = pubkeys.size
}

@Serializable
data class FollowingSnapshotStore(
    val snapshots: List<FollowingSnapshot>,
) {
    companion object {
        const val MAX_SNAPSHOTS = 25
    }
}

/**
 * A historical kind-3 contact list event discovered by scanning relays.
 * Used by Following Backup recovery to restore an older following list.
 */
data class Kind3Event(
    val id: String,
    val createdAt: Long, // unix seconds
    val pubkeys: List<String>,
    val pTags: List<List<String>>,
    val content: String,
) {
    val followCount: Int get() = pubkeys.size
}
